### ISSUE TITLE
Slow down when running on a large set of objects --- Memory leak?

### ISSUE THREAD

**Issue by JTerwel** [JTerwel/late-time_lc_binner#3]
After running this on the test objects, it seemed that doing everything took only a couple of seconds for each object.  I didn't time it though, this is just by eye.
When running it on a sample of 83 objects, the average time per object seemed to be around 10 seconds.
When I ran it on a sample of 653 objects, I expected it to take about 2 hours, seeing as each object would be taking about 10 seconds. It turned out to take an estimated 13 hours, or in other words ~70 seconds per object. This wasn't obvious for the first 100 or so, but as the program got further down the list, each new object took longer and longer.

What I think the issue is here, is that somehow, somewhere, not everything is deleted after the program is finished with an objects, and has to move to the next. This would cause things to start piling up somewhere in the memory, reducing the amount of memory available for use.

In any case, this is problematic. If this program is going to be tested on mock spectra, it will have to process a lot more that 653 objects, and waiting a week for it to finish is not the best idea. If the first 3 objects can be done in less than 5 seconds each, the last 3 objects should also take 5 seconds each.

So in short, program slows down. Why is this happening? Fix the issue once found!

**Comment by JTerwel** [comment]
Looking at the last object, ZTF18aajpjdi, there is about 1 minute between each plot made. The first plot made is in plot_lc(), then all the binning is done, and then plot_bin_results() is called 3 times, once for each filter. After that the bins and settings are saved.
Since the bins and settings are saved within the same minute as the last plot, there is probably no problem there (for now).

Instead, the issue seems to lie with the making of the plots, and my guess is that the issue is with not properly closing the plot. I had a similar issue last year when I didn't know this had to be done when making and saving plots in a loop. After that I started using plt.close() and I used it in here as well in both plot_lc() and plot_bin_result(). However, now I think that it doesn't close all subplots, or something like that, and the program still ends up holding onto all previously made plots.
See [this issue](https://github.com/matplotlib/matplotlib/issues/8560/) for reference.

No solution yet, but at the very least a hint to what could be the problem.

**Comment by JTerwel** [comment]
I did some testing:
Running the code on 8 objects: between ~ 6 - 9 s per object
Running the code on 83 objects: same time needed as before at start, but the average rose linearly to ~20 s at the last couple of objects

Next: tracing memory while running on 8 objects using tracemalloc
Result: top 5 memory consumers after all objects were finished were from matplotlib, using a total of ~ 175MiB
running tracemalloc also slowed down the program, but that is temporarily
adding plt.cla(), plt.clf() just before plt.close(), and gc.collect() to the end of the loop over each object, removed most of this memory. The top consumer now uses only 477KiB. Still not 0, but quite an improvement. It does slow down each object by ~ 0.5 s or so.
Running it again on 83 objects: time needed per objects at the start is ~ 6-9 s, time needed per object at the end is ~ 6-9 s
In other words, it seems to stay quite stable, which is quite the improvement.

[This question](https://stackoverflow.com/questions/7101404/how-can-i-release-memory-after-creating-matplotlib-figures) provided some good alternatives to look into, including multiprocessing.

The current state seems to work better / faster, but I'm convinced I can do better, so this issue will stay open for now. Also, the modified version of the code is not yet on github, I'll do that later as well.

**Comment by JTerwel** [comment]
I have taken a look at, and played a bit with, the multiprocessing module, here is what I found:
1) Not too difficult: After a bit of reading and trying, it is quite easy to set things up and have it work decently

2) Faster: This one should be obvious, but when I was testing the memory usage when making and saving 16 plots, it was very obvious to see how much faster things went when using multiprocessing instead of a loop

3) Less memory used: Even after just these 16 plots, there was less memory used with the multiprocessing. When just looping to make the plots, the top 9 items are matplotlib related, and the 10th one is bootstrap_external. With the multiprocessing, this bootstrap_external became the top memory user (with the same amount used as before), followed by multiprocessing things, but with less memory in use in total. Additionally, it didn't seem to matter whether or not I used cla(), clf(), and close(). It seems that all lingering matplotlib (spawned process) objects are remove automatically, which is nice and would work even better than the previous solution.

4) Changes needed: If I am to implement this into the main binning script, some large changes are needed. This includes, but doesn't limit to, the loop over all light curves (obviously, as this is where the multiprocessing would be implemented), how the main binning function is called (might have to zip values or reduce number of args), and the progress bar (I kind of want to keep it, as it is quite useful to see, but am not sure how to do that yet)

**Comment by JTerwel** [comment]
Multiprocessing was added, main function was changed to make this work, and the progress bar is now created using the tqdm package.

After some testing, this is what I found regarding to the speed of the program:
Using a list of 83 objects:
- Old way (6-9 s per object at the start, ~20 s per object at the end): ~ 18 minutes total
- Improved way by adding plt.cla(), plt.clf(), & gc.collect(): ~ 10 minutes total
- Multiprocessing (8 cpu cores): ~ 2.3 s per object, 188 s total (= just over 3 minutes) --> ~ 9x faster
- Multiprocessing without closing plots: ~ 2.3 s per object, 190 s total
- Multiprocessing without making plots: ~ 0.8 s per object, 63 s total --> ~ 18x faster

Using a list of 653 objects:
- Old way (6-9 s per object at the start, > 5 min per objet at the end): ~ 13 hours (estimated)
- Multiprocessing (8 cpu cores & using plt.clf() & plt.cla()): ~ 2.2 s per object, 23 min & 45 s total --> ~ 33x faster
- Multiprocessing without making plots: ~ 0.7 s per object, 7 min & 40 s total --> ~ 100x faster

So in short, this makes things a lot faster. One noteworthy side effect of trying to use the full capability of the computer is the energy consumption. During my trials, which lasted between 60 and 90 minutes total, I managed to drain the battery by ~ 85%. Also, while running, the fans will go spin at maximum speed to help cooling as much as possible. This happens more when making & saving plots, as expected. It may therefore be a good idea to keep this in mind & not run this without a good power supply.

With this the issue seems resolved and will be closed.
